### PR TITLE
Feature methods

### PR DESCRIPTION
**Pull request by srueda27**
A re-implementation of the #each, #each_with_index, #select, #all?, #any?, #none?, #count, #map, #inject methods of the Enumerable module from Ruby
The #my_map has a modification that can receives either a Proc or a block
You will find some testing that I made for every method inside the _spec file. The rake command will run the tests

### CODE REVIEW

File: enumerable.rb [hunk 1]
```diff
@@ -0,0 +1,170 @@
...
+      # return true unless any of the values does not meet the statement
+      if is_a? Hash
+        return false unless yield n.to_a[0], n.to_a[1]
+      elsif !(yield n)
+        return false
+      end
+    end
+
+    true
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`all?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-all-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_all.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true when none of the collection members are false or nil
- [x] when a class is passed as an argument returns true if all of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true if all of the collection matches the Regex

File: enumerable.rb [hunk 2]
```diff
@@ -0,0 +1,170 @@
...
+    my_each do |n|
+      if is_a? Hash
+        return true unless yield n.to_a[0], n.to_a[1]
+      elsif !(yield n)
+        return true
+      end
+    end
+
+    false
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`none?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_none.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true only if none of the collection members is true
- [x] when a class is passed as an argument returns true if none of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns true only if none of the collection matches the Regex
- [ ] when a pattern other than Regex or a Class is given returns true only if none of the collection matches the pattern

File: enumerable.rb [hunk 3]
```diff
@@ -0,0 +1,170 @@
...
+      # return true if any of the values meet the statement
+      if is_a? Hash
+        return true if yield n.to_a[0], n.to_a[1]
+      elsif yield n
+        return true
+      end
+    end
+
+    false
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`any?`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-any-3F). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_any.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] when no block or argument is given returns true if at least one of the collection is not false or nil
- [x] when a class is passed as an argument returns true if at least one of the collection is a member of such class
- [ ] when a Regex is passed as an argument returns false if none of the collection matches the Regex
- [x] when a pattern other than Regex or a Class is given returns false if none of the collection matches the pattern

File: enumerable.rb [hunk 4]
```diff
@@ -0,0 +1,170 @@
...
+    my_each do |n|
+      if is_a? Hash
+        count += 1 if yield n.to_a[0], n.to_a[1]
+      elsif yield n
+        count += 1
+      end
+    end
+
+    count
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`count`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-count). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_count.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [x] counts the number of items in enum that are equal to item if an argument is given

File: enumerable.rb [hunk 5]
```diff
@@ -0,0 +1,170 @@
...
+                              parameters_array[0]
+                            end
+    end
+
+    array.my_each do |n|
+      cumulative_variable = yield cumulative_variable, n
+    end
+
+    cumulative_variable
+  end
```
Comment (whiz25) [inline enumerable.rb]: - [ ] should be identical and returns the same thing as ruby's [`inject`](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject). ([Screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/ruby/advanced_building_blocks_enumerable/my_inject.png) from [Odin](https://www.theodinproject.com/courses/ruby-programming/lessons/advanced-building-blocks#assignment-2))
- [ ] combines all elements of enum by applying a binary operation, specified by a block
- [ ] when a symbol is specified combines each element of the collection by applying the symbol as a named method

**Review by whiz25 — CHANGES_REQUESTED** [review]
## Nice Job :+1: @srueda27 
Your methods cover most of the edge cases for this project. However, we need to fix a few issues before we can merge this PR.
- [ ] remove all comments from your code
- [ ] check my comments for suggestions
- [ ]  resubmit for another review after the changes

### Martin Afani

File: enumerable.rb (hunk 3 above)
Comment (srueda27) [reply enumerable.rb]: Could you please explain me better this one: "when a pattern other than Regex or a Class".
a Pattern isn't a Regex?

Comment (whiz25) [reply enumerable.rb]: ## Hey @srueda27 
What it means is that you need to test that method with a different patter than a regex.
For example, if we have `let(:words) { %w[dog door rod blade] }` as an array, we could try;
```ruby
expect(words.my_any?('cat')).to be words.any?('cat')
```
It should return `false` when none of the collection matches the pattern.

**Comment by Anna-Myzukina** [comment]
### Your project looks :+1:  good!
### Great job thus far. Let`s fix enumerables my_none? and enumerables my_inject  before it can be approved.

* should have working my_none?  and returns the same thing as ruby's none?.  You can read about  [none?](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-none-3F)

Now your my_none? method gives next errors:




* should have working my_inject and returns the same thing as ruby's inject. You can read about  [inject](https://ruby-doc.org/core-2.6.4/Enumerable.html#method-i-inject)

Now your inject my_method gives next errors:






### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*

**Review by Anna-Myzukina — CHANGES_REQUESTED** [review]
Needs to fix my_none? and my_inject methods

**Comment by Anna-Myzukina** [comment]
### Great job!
* But still has errors the same methods

er1
er2
er33


### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*

**Comment by Anna-Myzukina** [comment]
#### Great job! Just fix the next errors in enumerables my_inject and it can be approved.

* Your method my_inject gives next errors:
inject1
inject2


### Make changes and submit another code review request.

*If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you.*

**Comment by srueda27** [comment]
> #### Great job! Just fix the next errors in enumerables my_inject and it can be approved.
> 
>     * Your method my_inject gives next errors:
>       inject1
>       inject2
> 
> 
> ### Make changes and submit another code review request.
> 
> _If you have any questions, let me know on slack (@Anna-Muzykina) and I will be glad to clarify with you._

I'm running this tests and they are passing so, could you please hand me yours to see where my code is failing 
```
expect((1..10).my_inject(:+)).to eq((1..10).inject(:+))
expect([1,5,8,9,1,9].my_inject(:+)).to eq([1,5,8,9,1,9].inject(:+))
```

**Comment by shshamim63** [comment]
## Status :tada: Approved :tada: 
___
Hi @srueda27 :wave: ,
Code looks neat and clean.You have resolved all the issues raised by the previous TSEs. Which tends it to cover all the edge case scenarios. Repository contains a descriptive readme. Awesome job :clap: .
___
#### This :+1: from me. Best of luck for you next project.
___
## :herb: Happy Coding :herb: 
### Shakhawat Hossain
[Linkedin](https://www.linkedin.com/in/shakhawathossainshamim/)

**Pull request merged**